### PR TITLE
Make lookup constraints independent of choice of domain

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR tweaks the behaviour of `UnnormalizedLagrangeBasis` so that it can take positive or negative values (relative to the domain size). This allows us to make the lookup argument constraints independent of the domain size.

As a side-effect, this also gives us 'negative' powers that we can apply to the domain's generator. In the verifier circuit, this lets us perform a much smaller number of multiplications by calculating `1/g^(-1)`, rather that performing a large number of squarings.

### CODE REVIEW

**Review by jspada — APPROVED** [review]
Looks good to me.

File: kimchi/src/circuits/expr.rs [hunk 1]
```diff
@@ -475,9 +475,15 @@ impl<F: FftField> PolishToken<F> {
                 EndoCoefficient => stack.push(c.endo_coefficient),
                 Mds { row, col } => stack.push(c.mds[*row][*col]),
                 VanishesOnLast4Rows => stack.push(eval_vanishes_on_last_4_rows(d, pt)),
-                UnnormalizedLagrangeBasis(i) => stack.push(
-                    d.evaluate_vanishing_polynomial(pt) / (pt - d.group_gen.pow(&[*i as u64])),
-                ),
+                UnnormalizedLagrangeBasis(i) => {
+                    // Renormalize negative values to wrap around at domain size
+                    let i = if *i < 0 {
+                        ((*i as i64) + (d.size as i64)) as u64
```
Comment (mimoo) [inline kimchi/src/circuits/expr.rs:481]: this is great!

Comment (mimoo) [reply kimchi/src/circuits/expr.rs:481]: would still be nice to add tests related to this, off by ones scare me :)

Comment (jspada) [reply kimchi/src/circuits/expr.rs:481]: good idea

**Pull request merged**